erts/emulator/beam/erl_db.c

```diff
@@ -783,57 +807,57 @@ DbTable* db_get_table_aux(Process *p,
 
     ASSERT((what == DB_READ_TBL_STRUCT) == (kind == LCK_NOLOCK_ACCESS));
 
-    if (META_DB_LOCK_FREE())
-        name_already_locked = 1;
-
     if (is_not_atom(id)) {
         tb = tid2tab(id, &p->fvalue);
     } else {
-	struct meta_name_tab_entry* bucket = meta_name_tab_bucket(id,&name_lck);
-	if (!name_already_locked)
-	    erts_rwmtx_rlock(name_lck);
-	else {
-	    ERTS_LC_ASSERT(META_DB_LOCK_FREE()
-                           || erts_lc_rwmtx_is_rlocked(name_lck)
-                           || erts_lc_rwmtx_is_rwlocked(name_lck));
-            name_lck = NULL;
-	}
+        struct meta_name_tab_bucket* bucket = meta_name_tab_bucket(id);
+        struct meta_name_tab_entries *entries =
+            (struct meta_name_tab_entries*) erts_atomic_read_ddrb(&bucket->entries);
+        Uint bucket_size = erts_atomic32_read_rb(&entries->size);
+
         tb = NULL;
-	if (bucket->pu.tb != NULL) {
-	    if (is_atom(bucket->u.name_atom)) { /* single */
-		if (bucket->u.name_atom == id)
-		    tb = bucket->pu.tb;
-	    }
-	    else { /* multi */
-		Uint cnt = unsigned_val(bucket->u.mcnt);
-		Uint i;
-		for (i=0; i<cnt; i++) {
-		    if (bucket->pu.mvec[i].u.name_atom == id) {
-			tb = bucket->pu.mvec[i].pu.tb;
-			break;
-		    }
-		}
-	    }
-	}
 
-	if (tb == NULL) {
-            if (name_lck)
-                erts_rwmtx_runlock(name_lck);
+        for (Uint i = 0; i < bucket_size; ++i) {
+            if (entries->data[i].name_atom == id) {
+                tb = entries->data[i].tb;
+                break;
+            }
+        }
+
+        if (tb == NULL) {
             p->fvalue = EXI_ID;
-	}
+        }
     }
 
     if (tb) {
         if (what == DB_READ_TBL_STRUCT) {
-            if (name_lck)
-                erts_rwmtx_runlock(name_lck);
             return tb;
         }
```

Comment:
I think you must check the name for DB_READ_TBL_STRUCT as well. Otherwise we may start to operate on a completely random wrong table which feels a bit scary. At least I think it can result in incorrect errors if the random table has a different `keypos` for example.